Clean up IssuerFlow

rnicoll opened the pull request:
* Switch to using anonymous party as recipient
* Enable anonymisation for issuance as well as move in issuer flows.
* Pass notary into issuer flow rather than taking a notary at random from the network map.
* Enable anonymisation in IssuerFlowTest

josecoll commented:
I don't understand the requirement to provide anonymity in issuance.
From a coding perspective it feels wrong to be passing a boolean flag (`val anonymous: Boolean`) around and executing different logic based on this flag. Isn't the whole point of AnonymousParty to reflect the fact a party is anonymous? (so why a separate boolean flag).

Perhaps best to get a second reviewer on this item.

rnicoll commented:
So while we conventionally issue to ourselves, in theory you can issue directly to a third party, and therefore this anonymises the recipient (not the issuer). The flag is used to determine whether we create an anonymous identity for the transaction or not, because we may want to intentionally pay a well known identity, and right now some of the tests don't work with anonymisation enabled. The logic probably needs cleaning up, though, let me see if I can make it a bit clearer.

mikehearn commented:
Perhaps this is a good chance to simplify the cash app a bit. Do we really need to be able to directly issue to third parties? If so then there needs to be a separate flow to receive the issuance transaction. Can we just always insist on an issue+payment instead?

Likewise, being able to request that someone issues you cash via a flow doesn't make any sense does it. We should delete that feature.

rnicoll commented:
Corrected the notary service identity resolution in trader demo and rebased, and this should now pass CI.

`IssuerFlow` is used in a number of different tools and examples and we'd need to replace it with something else. Realistically that means a full trade which currency is then issued to fulfil. That's worth doing, but is beyond the scope of this PR.

rnicoll commented:
Rebased and ready for review

shamsasari commented on finance/src/test/kotlin/net/corda/flows/IssuerFlowTest.kt:
```diff
@@ -53,34 +53,71 @@ class IssuerFlowTest {
+                            require(update.produced.size == 1) { "Expected 1 produced states, actual: $update" }
+                        }
+                        // MOVE
+                        /* expect { update ->
```
Why is this commented out?

rnicoll replied:
The framework for checking transaction states hangs rather than reporting any useful error here. I'm guessing something in how I've used it, but as the same functionality is covered in an integration test I've removed this check and replaced it with a TODO to investigate further.

shamsasari commented on samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/api/BankOfCordaWebApi.kt:
```diff
@@ -43,23 +46,27 @@ class BankOfCordaWebApi(val rpc: CordaRPCOps) {
+        } catch (e: Exception) {
+            val out = StringWriter()
+            e.printStackTrace(PrintWriter(out))
+            logger.error("Issue request failed: ${e}")
```
Why we doing this and not simply `logger.error("Issue request failed", e)`

rnicoll replied:
Old debugging code that's been left in, and had specifically wanted it to dump to stdout at the time. I'll switch that to logging the exception though, which it should have done in the first place.

shamsasari commented on finance/src/test/kotlin/net/corda/flows/IssuerFlowTest.kt:
```diff
@@ -90,14 +127,15 @@ class IssuerFlowTest {
-        val issueRequest = IssuanceRequester(amount, party, issueToPartyAndRef.reference, issuerNode.info.legalIdentity,
-                anonymous = false)
+        val issueRequest = IssuanceRequester(amount, issueToParty, issueToPartyAndRef.reference, issuerNode.info.legalIdentity,
+                notaryParty, anonymous = true)
```
So this is the key bit of the test. From what I understand issuance can be both anonymous and non-anonymous? Should this test class be renamed to AnonymousIssuerFlowTest?

rnicoll replied:
I've now added new tests covering version using both anonymous and well known identities, now, and rebased.

josecoll commented on finance/src/test/kotlin/net/corda/flows/IssuerFlowTest.kt:
```diff
@@ -53,34 +53,95 @@ class IssuerFlowTest {
+                            require(update.produced.size == 1) { "Expected 1 produced states, actual: $update" }
+                        }
+                        // TODO: Should check the move transaction consumes one input and produces zero, however
+                        // for some reason the "expect" hangs if we try. Verifying on the receiving side is sufficient,
```
This should be investigated further before Approval as it is **functionally incorrect** if the sender is not receiving the spend Update.

rnicoll replied:
I've pulled the issuance vault tests out into their own PR #995 so they can be corrected in isolation from the functional changes.

shamsasari commented on finance/src/test/kotlin/net/corda/flows/IssuerFlowTest.kt:
```diff
@@ -53,34 +53,95 @@ class IssuerFlowTest {
     }
 
     @Test
+    fun `test issuer flow without anonymisation`() {
```
Can you use `JunitParams` https://github.com/Pragmatists/JUnitParams. It was written just for cases like this.

rnicoll commented:
Rebased after the underlying text issue is fixed, this is ready for re-review now

josecoll commented:
As mentioned previously, I don't feel we should be passing a Boolean flag around called `anonymous`. Please remove me from reviewers list.

rnicoll commented:
@mikehearn Substantial changes to the issuer flow (especially removing it) would impact several of the samples, and the demo bench I believe. I'd rather tackle that in a different PR later, and get the identity work progressed meanwhile.

mikehearn commented:
I don't think it implies removing the issuer flow, just making sure it's not marked as InitiatedBy so other peers can't request that we give them money. But OK to tweaking the app in a separate PR.
